README
---------------------------------------------------------------------------

                            Distribution of RES
            Automatically Resizing Contiguous Memory for OCaml

---------------------------------------------------------------------------

                               Prerequisites

             YOU WILL NEED GNU-MAKE TO COMPILE THE SYSTEM WITH
                         THE DISTRIBUTED MAKEFILES

---------------------------------------------------------------------------

                       Contents of this distribution

Changes             - History of code changes.

INSTALL             - Short notes on compiling and installing the library

LICENSE             - A copy of the "GNU LESSER GENERAL PUBLIC LICENSE"

Makefile            - Top Makefile

OcamlMakefile       - Makefile for easy handling of compilation of not
                      so easy OCaml-projects.  It generates dependencies
                      of Ocaml-files automatically, is able to handle
                      "ocamllex"-, "ocamlyacc"-, IDL- and C-files and
                      generates native- or byte-code, as executable or
                      as library - with thread-support if you want!

README              - this file

examples/

lib/                - Implementation of the RES-library.

---------------------------------------------------------------------------

                                What is it?

This OCaml-library consists of a set of modules which implement
automatically resizing (= reallocating) datastructures that consume
a contiguous part of memory. This allows appending and removing of
elements to/from arrays (both boxed and unboxed), strings (-> buffers),
bit strings and weak arrays while still maintaining fast constant-time
access to elements.

There are also functors that allow the generation of similar modules
which use different reallocation strategies.

---------------------------------------------------------------------------

                      Why should you use it?

For several reasons:

  * Fast constant-time access to indexed elements (e.g. in arrays and
    strings) is often a prerequisite for short execution times of
    programs.

    Still, operations like adding and/or removing elements to/from the
    end of such datastructures are often needed.  Unfortunately, having
    both properties at the same time sometimes requires reallocating
    this contiguous part of memory.

    This module does not eliminate this problem, but hides the process
    of reallocation from the user, i.e. it happens automatically.

    Thus, the user is liberated from this bug-attracting (e.g. index
    errors) task.

  * This library allows the user to parameterize allocation strategies at
    runtime. This is a very important feature, because it is impossible
    for any allocation algorithm to perform optimally without having
    knowledge about the user program.

    For example, the programmer might know that a consecutive series
    of operations will alternately add and remove large amounts of
    elements. In such a case it would be wise to keep a high reserve
    of available slots in the datastructure, because otherwise it will
    resize very often during this procedure which requires a significant
    amount of time.

    By raising a corresponding threshold in appropriate places at runtime,
    the programmer can fine-tune the behaviour of e.g. his buffers for
    optimal performance and set this parameter back later to save memory.

  * Because reallocation strategies themselves may be quite complicated,
    it was also a design goal to have the user supply his own ones
    (if required).

    By using functors the user can parameterize these datastructures
    with his own reallocation strategies, giving him even more control
    over how and when reallocations are triggered.

  * It is possible that the user wants to add support for additional
    low-level implementations that require reallocations. Even in this
    case it is fairly easy to create new modules by using functors.

  * The library implements a large interface of functions, all of which
    are completely independent of the reallocation strategy and the
    low-level implementation.

    All the interfaces of the corresponding low-level implementations
    of datastructures (e.g. array, string) are fully supported and have
    been extended with further functionality. There is even a new buffer
    module which can be used in every context of the standard one.

  * OCaml makes a distinction between unboxed and boxed arrays. If
    the type of an array is "float", the representation will be unboxed
    in cases in which the array is not used in a polymorphic context
    (native code only).

    To benefit from these much faster representations there are
    specialized versions of automatically resizing arrays in the
    distribution.

---------------------------------------------------------------------------

                        Documentation of Functions

The preparameterized modules (default strategy) and the functors for
mapping strategy-implementations to this kind of modules are contained
and documented in file "lib/res.mli".

For examples of how to use the functors to implement new strategies and/or
low-level representations, take a look at the implementation of "res.ml".

Their function interface, however, is documented in files
"lib/pres_intf.mli" (for parameterized "low-level" types like e.g. normal
arrays) and in "lib/nopres_intf.mli" (for non-parameterized "low-level"
types like e.g. int arrays, strings (buffers), ...).

This was done so as not do confuse people with too large an interface
description in one file.

It is advisable to keep all of these files at some specific location at
hand for quick information.

---------------------------------------------------------------------------

                          Hints for Convenient Use

It should be noted that it is possible to use the standard notation for
accessing elements (e.g. "ar.(42)") with resizable arrays (and even with
"Buffer", "Bits", etc...). This requires a short explanation of how OCaml
treats such syntactic sugar:

All that OCaml does is that it replaces each such structure to an appropriate
"Array.get" or "Array.set". This may be *any* module that happens to be bound
to this name in the current scope! The same principle is true for the
String-module and the '.[]'-operator.

Thus, the following works:

  module Array = Res.Bits
  module String = Res.Buffer

  let _ =
    let ar = Array.empty () in
    Array.add_one ar true;
    print_endline (string_of_bool ar.(0));
    let str = String.empty () in
    String.add_one str 'x';
    print_char str.[0]; print_newline ()

Do not forget that it is even possible to open modules locally! Example:

  let _ =
    let module Array = Res.Array in
    Printf.printf "%d\n" (Array.init 10 (fun x -> x * x)).(7)


If you want to change one of your files to make use of resizable arrays
instead of standard ones without much trouble, you should read the
following:

You should "save" the standard Array-module and its type for later access:

  module StdArray = Array
  type 'a std_array = 'a array

Make the resizable implementation (includes the index operators!)
available:

  open Res

or explicitely:

  module Array = Res.Array

or if you want to use a specific "Array"-implementation:

  module Array = Res.Bits

Then set the type:

  type 'a array = 'a Array.t

If you create standard arrays with the builtin syntax, change lines like

  let ar = [| 1; 2; 3; 4 |] in

to

  let ar = Array.of_array [| 1; 2; 3; 4 |] in

This should allow all of your sources to compile out-of-the-box with the
additional functionality. In places where you still need the standard
implementation you should have no problems to use the rebound module
and type to do so.

This trick works similarly for the old and the new Buffer-module. You
might also want to replace the "String" module in this fashion. The
latter one, however, supports a number of functions like e.g. "escape",
which are not available then.

---------------------------------------------------------------------------

Up-to-date information (newest release of distribution) can always be
found at:

  http://www.ocaml.info/home/ocaml_sources.html

---------------------------------------------------------------------------

Enjoy!

Vienna, 2002-09-11
Markus Mottl

e-mail: markus.mottl@gmail.com
WWW:    http://www.ocaml.info


Changes
2008-09-16: Changed strategy API to greatly improve performance of
            growing/shrinking.

2008-05-09: Added unsafe_expose_array to parameterized resizable arrays.

2006-11-22: Updated OCamlMakefile.

2005-12-26: Fixed a build problem.

2005-10-24: Added sof_list.

2004-04-11: Removed use of unsafe external function that depends on
            current CVS-version.

2004-01-28: Renamed external function for compatibility with most recent
            OCaml-version.

            Updated OCamlMakefile.

2003-04-09: Updated OCamlMakefile.
            Fixed an installation problem.

2003-01-07: Updated OCamlMakefile to make use of "findlib".

2002-09-23: Fixed a bug in "remove_n" (arguments not fully checked).

            Slightly improved efficiency.

2002-09-11: Updated OCamlMakefile and license.

            Documented all modules for ocamldoc.

            Changed module Res for better accessibility.

            Made resizable weak arrays conform to module Weak again.

2002-05-04: Revised the whole installation procedure. See INSTALL for
            details.

2002-04-30: Updated OcamlMakefile: it does not ask for confirmation
            during installation anymore!

2001-06-30: Removed "Printexc.catch" from stupid_ga-example: is going
            to be deprecated in upcoming OCaml-release.

2001-06-24: Added special module for resizable integer arrays (again),
            because it is faster on many operations.

2001-01-30: Made Makefile more general (allows simpler addition of
            further examples).

2001-01-26: Made use of the new OCaml-keyword "include" for module
            inclusion. This makes the file "lib/res.ml" significantly
            shorter. This change requires an OCaml-version higher
            than 3.00.

2001-01-24: Updated OcamlMakefile

2000-06-24: Updated OcamlMakefile

2000-06-13: Updated OcamlMakefile

2000-06-11: Updated OcamlMakefile

2000-06-08: Added installation routine + updated OcamlMakefile again:

            This upgrade makes installation much easier! Read the
            updated INSTALL-file!

2000-06-07: Upgraded to new OcamlMakefile.

2000-04-28: Fixed *critical* bug:

            Filling and blitting accidently truncated the array if the last
            index of the operation was smaller than the one of the target
            array.

            Resizable bit-vectors should be *much* more efficient now
            (blitting, resizing, etc. about 30 (60) times faster, depending on
            your architecture!):

            I took the new implementation of Jean-Christophe Filliatre's
            bitv-library, which uses some very clever algorithms for efficient
            blitting. In the near (?) future I'll also add his functions for
            common logical, efficient operations on bit-strings (unless
            somebody wants to volunteer... ;-)

2000-03-23: Removed special module for resizable integer arrays:

            Integer arrays are not unboxed and won't be in the (near?)
            future: this would cause generic polymorphic functions such as
            equality, hashing and output_value to produce wrong results.

            Therefore, use the parameterized version instead. It is
            equally fast.

2000-03-08: New function (in all implementations):

            find_index - takes a predicate, a resizable array and a
                         start index and returns the index of the
                         first element that satisfies the predicate -
                         see interface documentation for details.

            Fixed documentation of interfaces: in some cases the wrong
            name for possibly raised exceptions was provided.

2000-01-10: Added functions for converting standard arrays to resizable
            ones and strings to buffers.

            Added "create" and "screate" to the interface of parameterized
            arrays. This makes it easier to use it in place of the
            standard array.

            Removed "make" and "smake" from resizable weak arrays -
            not useful there.

            Updated documentation on how to use the index operators with
            the resizable datastructures and how to easily replace the
            standard arrays/strings with the resizable ones in large
            sources.

1999-12-25: Added support for weak arrays + small example

1999-11-04: Added support for bit-vectors
            (peeked at Jean-Christophe Filliatre's bitv-library for this).

            Added new example:

              stupid_ga.ml (a brain-dead genetic algorithm using bit-vectors)

            We now have a TODO-list ;-)

1999-10-23: Added three new functions:

            remove_range - removes a range of elements within a resizable
                           array

            pos  - returns the index of the first logically equal element
            posq - returns the index of the first physically equal element

1999-10-13: First release.


lib/pres_impl.ml
(*
   RES - Automatically Resizing Contiguous Memory for OCaml

   Copyright (C) 1999-2002  Markus Mottl
   email: markus.mottl@gmail.com
   WWW:   http://www.ocaml.info

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*)

(* $Id: pres_impl.ml,v 1.16 2005/11/07 20:25:28 mottl Exp $ *)

module type Implementation = sig
  type 'a t
  val name : string
  val length : 'a t -> int
  val make : int -> 'a -> 'a t
  val unsafe_get : 'a t -> int -> 'a
  val unsafe_set : 'a t -> int -> 'a -> unit
end

module Make (S : Strat.T) (Impl : Implementation) = struct
  module Strategy = S

  type strategy = Strategy.t

  type 'a t = {mutable ar : 'a Impl.t;
               mutable vlix : int;
               mutable strategy : strategy}

  let name = Impl.name

  let invalid_arg str = invalid_arg (name ^ "." ^ str)
  let failwith str = failwith (name ^ "." ^ str)

  let length ra = succ ra.vlix
  let lix ra = ra.vlix 

  let real_length ra = Impl.length ra.ar
  let real_lix ra = pred (real_length ra)

  let unsafe_get ra ix = Impl.unsafe_get ra.ar ix
  let unsafe_set ra ix el = Impl.unsafe_set ra.ar ix el

  let unsafe_expose_array ra = ra.ar

  let get ra n =
    if n > ra.vlix or n < 0 then invalid_arg "get" else unsafe_get ra n

  let set ra n el =
    if n > ra.vlix or n < 0 then invalid_arg "set" else unsafe_set ra n el

  let dummy_loc = 0
  let no_obj () = Obj.magic dummy_loc

  let creator n = let nothing = no_obj () in Impl.make n nothing

  let screate_fresh strategy n =
    let res = {ar = no_obj (); vlix = pred n; strategy = strategy} in
    res.ar <- creator (Strategy.grow strategy  n);
    res

  let create_fresh n = screate_fresh Strategy.default n

  let create_from ra =
    {ar = creator (length ra); vlix = ra.vlix; strategy = ra.strategy}

  let sempty strategy =
    let res = {ar = no_obj (); vlix = -1; strategy = strategy} in
    res.ar <- creator (Strategy.grow strategy 0);
    res

  let empty () = sempty Strategy.default

  let screate strategy n x =
    let res = screate_fresh strategy n in
    for i = 0 to pred n do unsafe_set res i x done; res

  let smake = screate

  let create n = smake Strategy.default n

  let make = create

  let sinit strategy n f =
    let res = screate_fresh strategy n in
    for i = 0 to pred n do unsafe_set res i (f i) done; res

  let init n f = sinit Strategy.default n f

  let to_array ra = Array.init (length ra) (fun i -> unsafe_get ra i)

  let sof_array strategy ar =
    sinit strategy (Array.length ar) (fun i -> Array.unsafe_get ar i)

  let of_array ar = sof_array Strategy.default ar

  let get_strategy ra = ra.strategy

  let resizer some_lix ra len =
    let ar = creator len in
    for i = 0 to some_lix do Impl.unsafe_set ar i (unsafe_get ra i) done;
    ra.ar <- ar

  let enforce_strategy ra =
    let real_len = real_length ra and new_len = length ra in
    let new_real_len = Strategy.shrink ra.strategy ~real_len ~new_len in
    if new_real_len <> -1 then resizer ra.vlix ra new_real_len

  let set_strategy ra strategy = ra.strategy <- strategy; enforce_strategy ra
  let put_strategy ra strategy = ra.strategy <- strategy

  let make_matrix sx sy init =
    let res = create_fresh sx in
    for i = 0 to res.vlix do unsafe_set res i (make sy init) done; res

  let copy ra =
    let ar = Impl.make (real_length ra) (unsafe_get ra 0) in
    for i = 1 to real_lix ra do Impl.unsafe_set ar i (unsafe_get ra i) done;
    {ra with ar = ar}

  let unsafe_blit_on_other ra1 ofs1 ra2 ofs2 len =
    let ofs_diff = ofs2 - ofs1 in
    for i = ofs1 to ofs1 + pred len do
      unsafe_set ra2 (i + ofs_diff) (unsafe_get ra1 i) done

  let append ra1 ra2 = match ra1.vlix, ra2.vlix with
    | -1, -1 -> empty ()
    | _, -1 -> copy ra1
    | -1, _ -> copy ra2
    | _ -> let len1 = length ra1
           and len2 = length ra2 in
           let res = create_fresh (len1 + len2) in
           unsafe_blit_on_other ra1 0 res 0 len1;
           unsafe_blit_on_other ra2 0 res len1 len2; res

  let rec concat_aux res offset = function
    | [] -> res
    | h::t -> if h.vlix < 0 then concat_aux res offset t
              else let len = length h in
                   unsafe_blit_on_other h 0 res offset len;
                   concat_aux res (offset + len) t

  let concat l =
    let len = List.fold_left (fun a el -> a + length el) 0 l in
    if len = 0 then empty ()
    else concat_aux (create_fresh len) 0 l

  let unsafe_sub ra ofs len =
    let res = create_fresh len in unsafe_blit_on_other ra ofs res 0 len; res

  let sub ra ofs len =
    if ofs < 0 or len < 0 or ofs + len > length ra then invalid_arg "sub"
    else unsafe_sub ra ofs len

  let guarantee_ix ra ix =
    if real_lix ra < ix then
      resizer ra.vlix ra (Strategy.grow ra.strategy (succ ix))

  let maybe_grow_ix ra new_lix = guarantee_ix ra new_lix; ra.vlix <- new_lix

  let add_one ra x = let n = length ra in maybe_grow_ix ra n; unsafe_set ra n x

  let unsafe_remove_one ra =
    unsafe_set ra ra.vlix (no_obj ()); ra.vlix <- pred ra.vlix;
    enforce_strategy ra

  let remove_one ra =
    if ra.vlix < 0 then failwith "remove_one" else unsafe_remove_one ra

  let unsafe_remove_n ra n =
    let old_vlix = ra.vlix and old_ar = ra.ar in
    ra.vlix <- old_vlix - n; enforce_strategy ra;
    if old_ar == ra.ar then
      let nothing = no_obj () in
      for i = succ ra.vlix to old_vlix do unsafe_set ra i nothing done

  let remove_n ra n =
    if n > length ra || n < 0 then invalid_arg "remove_n"
    else unsafe_remove_n ra n

  let unsafe_remove_range ra ofs len =
    unsafe_blit_on_other ra (ofs + len) ra ofs (length ra - len);
    unsafe_remove_n ra len

  let remove_range ra ofs len =
    if ofs < 0 or len < 0 or ofs + len > length ra then
      invalid_arg "remove_range"
    else unsafe_remove_range ra ofs len

  let clear ra = unsafe_remove_n ra (length ra)

  let unsafe_swap ra n m =
    let tmp = unsafe_get ra n in
    unsafe_set ra n (unsafe_get ra m);
    unsafe_set ra m tmp

  let swap ra n m =
    if n > ra.vlix or m > ra.vlix or n < 0 or m < 0 then invalid_arg "swap"
    else unsafe_swap ra n m

  let unsafe_swap_in_last ra n =
    unsafe_set ra n (unsafe_get ra ra.vlix);
    unsafe_remove_one ra

  let swap_in_last ra n =
    if n > ra.vlix or n < 0 then invalid_arg "swap_in_last"
    else unsafe_swap_in_last ra n

  let unsafe_fill ra ofs len x =
    let last = ofs + pred len in
    guarantee_ix ra (max last ra.vlix);
    for i = ofs to last do unsafe_set ra i x done

  let fill ra ofs len x =
    if ofs < 0 or len < 0 or ofs > length ra then invalid_arg "fill"
    else unsafe_fill ra ofs len x

  let unsafe_blit ra1 ofs1 ra2 ofs2 len =
    guarantee_ix ra2 (ofs2 + pred len);
    if ofs1 < ofs2 then
      for i = pred len downto 0 do
        unsafe_set ra2 (ofs2 + i) (unsafe_get ra1 (ofs1 + i)) done
    else
      for i = 0 to pred len do
        unsafe_set ra2 (ofs2 + i) (unsafe_get ra1 (ofs1 + i)) done

  let blit ra1 ofs1 ra2 ofs2 len =
    if len < 0 or ofs1 < 0 or ofs2 < 0 or ofs1 + len > length ra1 or
       ofs2 > length ra2 then invalid_arg "blit"
    else unsafe_blit ra1 ofs1 ra2 ofs2 len

  let rec to_list_aux ra i accu =
    if i < 0 then accu else to_list_aux ra (pred i) (unsafe_get ra i :: accu)

  let to_list ra = to_list_aux ra ra.vlix []

  let rec of_list_aux res i = function
    | [] -> res 
    | h::t -> unsafe_set res i h; of_list_aux res (succ i) t

  let of_list l = of_list_aux (create_fresh (List.length l)) 0 l

  let sof_list s l = of_list_aux (screate_fresh s (List.length l)) 0 l

  let iter f ra = for i = 0 to ra.vlix do f (unsafe_get ra i) done

  let map f ra =
    let res = create_from ra in
    for i = 0 to res.vlix do unsafe_set res i (f (unsafe_get ra i)) done; res

  let iteri f ra = for i = 0 to ra.vlix do f i (unsafe_get ra i) done

  let mapi f ra =
    let res = create_from ra in
    for i = 0 to res.vlix do unsafe_set res i (f i (unsafe_get ra i)) done; res

  let fold_left f accu ra =
    let res = ref accu in
    for i = 0 to ra.vlix do res := f !res (unsafe_get ra i) done; !res

  let fold_right f ra accu =
    let res = ref accu in
    for i = ra.vlix downto 0 do res := f (unsafe_get ra i) !res done; !res

  let rec for_all_aux i p ra =
    if i > ra.vlix then true
    else if p (unsafe_get ra i) then for_all_aux (succ i) p ra else false

  let for_all p ra = for_all_aux 0 p ra

  let rec exists_aux i p ra =
    if i > ra.vlix then false
    else if p (unsafe_get ra i) then true else exists_aux (succ i) p ra

  let exists p ra = exists_aux 0 p ra

  let rec mem_aux i x ra =
    if i > ra.vlix then false
    else if unsafe_get ra i = x then true else mem_aux (succ i) x ra

  let mem x ra = mem_aux 0 x ra

  let rec memq_aux i x ra =
    if i > ra.vlix then false
    else if unsafe_get ra i == x then true else memq_aux (succ i) x ra

  let memq x ra = memq_aux 0 x ra

  let rec pos_aux i x ra =
    if i > ra.vlix then None
    else if unsafe_get ra i = x then Some i else pos_aux (succ i) x ra

  let pos x ra = pos_aux 0 x ra

  let rec posq_aux i x ra =
    if i > ra.vlix then None
    else if unsafe_get ra i == x then Some i else posq_aux (succ i) x ra

  let posq x ra = posq_aux 0 x ra

  let rec find_aux i p ra =
    if i > ra.vlix then raise Not_found
    else let el = unsafe_get ra i in if p el then el else find_aux (succ i) p ra

  let find p ra = find_aux 0 p ra

  let rec find_index_aux p ra i =
    if i > ra.vlix then raise Not_found
    else if p (unsafe_get ra i) then i else find_index_aux p ra (succ i)

  let find_index p ra i =
    if i < 0 then invalid_arg "find_index" else find_index_aux p ra i

  let filter p ra =
    let res = sempty ra.strategy in
    for i = 0 to ra.vlix do
      let el = unsafe_get ra i in if p el then add_one res el done;
    res

  let find_all = filter

  let filter_in_place p ra =
    let dest = ref 0
    and pos = ref 0 in
    while !pos <= ra.vlix do
      let el = unsafe_get ra !pos in
      if p el then begin unsafe_set ra !dest el; incr dest end;
      incr pos done;
    unsafe_remove_n ra (!pos - !dest)

  let partition p ra =
    let res1, res2 as res = sempty ra.strategy, sempty ra.strategy in
    for i = 0 to ra.vlix do
      let el = unsafe_get ra i in
      if p el then add_one res1 el else add_one res2 el done;
    res
end
